### DISCUSSION TITLE
Configuring KEM groups in TLS1.3

### DISCUSSION THREAD

**Discussion by catharsis71** [open-quantum-safe/oqs-provider#252] [Q&A]
Hello, I recently started using this and I have a lot of questions.  Ubuntu 20.04.3, multiple servers, 1 ARM64 and the rest AMD64, OpenSSL 3.0.2, Apache 2.4.52, ntpsec (latest master build), Postfix 3.6.4 (no inbound connections, outbound to a specific SMTP server only).  All personal/hobby stuff so I don't mind running experimental software.

1.  In my ```openssl.cnf``` I did not have a pre-existing ```Groups =``` entry

I'm using this as a placeholder and it seems to work okay:

```Groups = x25519_kyber512:x25519_kyber768:sect571r1:sect571k1:secp521r1:brainpoolP512r1:sect409r1:sect409k1:brainpoolP384r1:secp384r1:X448:x25519:ffdhe8192:ffdhe6144:ffdhe4096:ffdhe3072:secp256r1```

but what I'd really like to do is tell it to add x25519_kyber512 and x25519_kyber768 on to the front of whatever the "default" group list is but I'm not sure how to do that

or if there's some kind of "all" or "default" tag like there is for the cypher list, maybe I could do something like that

this is mainly for curl since Apache has its own (more restrictive) group list.

2.  ```curl --curve x25519_kyber768``` works for connecting to my Apache servers, as well as https://pq.cloudflareresearch.com/  and random Cloudflare-proxied websites.    ```curl --curve x25519_kyber512``` works for connecting to my Apache servers, but does NOT work for connecting to https://pq.cloudflareresearch.com/ or random Cloudflare-proxied websites ("sslv3 alert handshake failure"), even though they claim to support it.

3.  ```curl --curve x25519_kyber768``` works for connecting to test.openquantumsafe.org port 6041 (ecdsap256/x25519_kyber768) and port 6084 (rsa3072/x25519_kyber768).  However, it cannot connect to the 'wildcard' port 6000 (ecdsap256/\*) or port 6043 (rsa3072/\*) - "sslv3 alert handshake failure"

Likewise ```curl --curve x25519_kyber512`` can connect to ports 6038 and 6081 but not 6000 or 6043.

4.  When I was doing initial Apache testing, I briefly experimented with configuring ```SSLOpenSSLConfCmd Curves``` with either x25519_kyber512 or x25519_kyber768 as the only curve.    With x25519_kyber768 as the only curve, Chromium browsers with Kyber support enabled can connect to my server, however, with x25519_kyber512 as the only curve, they can not (ERR_SSL_VERSION_OR_CIPHER_MISMATCH)

5.  Based on the above few items it seems like x25519_kyber512 as implemented by OQS, and x25519_kyber512 as implemented by Cloudflare and Chromium, are just not compatible with each other?  But Chromium and Cloudflare are compatible with each other, and of course OQS is compatible with itself.  But for x25519_kyber768, they are all compatible?

6.  After initial testing, I ended up with this as my "permanent" Apache configuration (for now):

```
SSLOpenSSLConfCmd Curves x25519_kyber512:x25519_kyber768:sect571r1:sect571k1:secp521r1:brainpoolP512r1:sect409r1:sect409k1:brainpoolP384r1:secp384r1:X448
```

this is in my global Apache configuration; don't worry about the kinda weird curve list, these are all the ones that the SSL Labs scan doesn't penalize for

all seems to work fine in terms of connectivity (except for Chromium not wanting to use x25519_kyber512 as noted above but it just uses 768 instead so it's fine)

my next though was to create PQ-only vhosts

I created one vhost with ```SSLOpenSSLConfCmd Curves x25519_kyber512:x25519_kyber768```
and then one with only x25519_kyber512 and one with only x25519_kyber768

according to the Apache docs, ```SSLOpenSSLConfCmd``` ***can*** be placed in a vhost but it's not actually working for me so far

it's ignoring the vhost curve configuration and using the global ```SSLOpenSSLConfCmd Curves``` even for these VLANs

I haven't made much headway on this yet

my only thought on what to try next is to remove the ```SSLOpenSSLConfCmd Curves``` from global Apache configuration and copy it into every single one of my normal vhosts.... but I have a lot of them so I haven't tried it yet

7.  ntpsec / NTS

I restarted ntpsec on all my servers and watched for breakages

seems like the only thing that broke was NTS to time.cloudflare.com

all the other NTS servers mine communicate with are probably falling back to normal X25519 or something else, as expected

but Cloudflare's NTS servers seem to be killing the connection when they see Kyber on offer

this is probably something I should ask Cloudflare about and I intend to try but I figured I should ask here first in case anybody knows anything

ntpsec log:

```
2023-09-09T07:07:58 ntpd[2050142]: DNS: dns_probe: time.cloudflare.com, cast_flags:1, flags:21901
2023-09-09T07:07:58 ntpd[2050142]: NTSc: DNS lookup of time.cloudflare.com took 0.044 sec
2023-09-09T07:07:58 ntpd[2050142]: NTSc: connecting to time.cloudflare.com:4460 => [2606:4700:f1::1]:4460
2023-09-09T07:07:58 ntpd[2050142]: NTSc: set cert host: time.cloudflare.com
2023-09-09T07:07:58 ntpd[2050142]: NTSc: SSL_connect failed
2023-09-09T07:07:58 ntpd[2050142]: NTS: error:0A0003E7:SSL routines::invalid session id
2023-09-09T07:07:58 ntpd[2050142]: NTSc: NTS-KE req to time.cloudflare.com took 0.048 sec, fail
2023-09-09T07:07:58 ntpd[2050142]: DNS: dns_check: processing time.cloudflare.com, 1, 21901
2023-09-09T07:07:58 ntpd[2050142]: DNS: dns_take_status: time.cloudflare.com=>error, 12
```

manual testing

as a reminder, I have this in openssl.cnf:

```
Groups = x25519_kyber512:x25519_kyber768:sect571r1:sect571k1:secp521r1:brainpoolP512r1:sect409r1:sect409k1:brainpoolP384r1:secp384r1:X448:x25519:ffdhe8192:ffdhe6144:ffdhe4096:ffdhe3072:secp256r1
```

Trying to connect to time.cloudflare.com port 4460 (NTS) without forcing a surve:

```
# openssl s_client -connect time.cloudflare.com:4460
CONNECTED(00000003)
4017864B5F7F0000:error:0A0003E7:SSL routines:tls_process_server_hello:invalid session id:../ssl/statem/statem_clnt.c:1467:
---
no peer certificate available
---
No client certificate CA names sent
---
SSL handshake has read 61 bytes and written 1142 bytes
Verification: OK
---
New, (NONE), Cipher is (NONE)
Secure Renegotiation IS NOT supported
Compression: NONE
Expansion: NONE
No ALPN negotiated
Early data was not sent
Verify return code: 0 (ok)
---
```

Forcing x25519_kyber512:

```
# openssl s_client -curves x25519_kyber512 -connect time.cloudflare.com:4460
CONNECTED(00000003)
4047F117AE7F0000:error:0A000126:SSL routines:ssl3_read_n:unexpected eof while reading:../ssl/record/rec_layer_s3.c:308:
---
no peer certificate available
---
No client certificate CA names sent
---
SSL handshake has read 0 bytes and written 1102 bytes
Verification: OK
---
New, (NONE), Cipher is (NONE)
Secure Renegotiation IS NOT supported
Compression: NONE
Expansion: NONE
No ALPN negotiated
Early data was not sent
Verify return code: 0 (ok)
---
```

Forcing x25519_kyber768:

```
# openssl s_client -curves x25519_kyber768 -connect time.cloudflare.com:4460
CONNECTED(00000003)
40C718EA897F0000:error:0A000126:SSL routines:ssl3_read_n:unexpected eof while reading:../ssl/record/rec_layer_s3.c:308:
---
no peer certificate available
---
No client certificate CA names sent
---
SSL handshake has read 0 bytes and written 1486 bytes
Verification: OK
---
New, (NONE), Cipher is (NONE)
Secure Renegotiation IS NOT supported
Compression: NONE
Expansion: NONE
No ALPN negotiated
Early data was not sent
Verify return code: 0 (ok)
---
```

But if I force normal X25519 it works:

```
# openssl s_client -curves x25519  -connect time.cloudflare.com:4460
CONNECTED(00000003)
(certificate stuff snipped)
---
Server certificate
-----BEGIN CERTIFICATE-----
(snipped)
-----END CERTIFICATE-----
subject=C = US, ST = California, L = San Francisco, O = "Cloudflare, Inc.", CN = time.cloudflare.com
issuer=C = US, O = DigiCert Inc, CN = DigiCert Global G2 TLS RSA SHA256 2020 CA1
---
No client certificate CA names sent
Peer signing digest: SHA256
Peer signature type: ECDSA
Server Temp Key: X25519, 253 bits
---
SSL handshake has read 3174 bytes and written 383 bytes
Verification: OK
---
New, TLSv1.3, Cipher is TLS_AES_256_GCM_SHA384
Server public key is 256 bit
Secure Renegotiation IS NOT supported
Compression: NONE
Expansion: NONE
No ALPN negotiated
Early data was not sent
Verify return code: 0 (ok)
---
---
Post-Handshake New Session Ticket arrived:
SSL-Session:
    Protocol  : TLSv1.3
    Cipher    : TLS_AES_256_GCM_SHA384
    Session-ID: E2BF14A068E79A21789CA6A0DE616CF1F6D9A5B1822F1B71B0A6B392E7F50BCC
    Session-ID-ctx:
    Resumption PSK: 5AD06A4DA4D44B64FB7A0A5FAD17BE50CED251CF2D38686E58064CEB92C0F6F1EC4ACD41DF164414E58130007FD62929
    PSK identity: None
    PSK identity hint: None
    SRP username: None
    TLS session ticket lifetime hint: 86400 (seconds)
    TLS session ticket:
    0000 - 3c 80 4b d1 d7 46 0c 91-cb c5 2a e4 eb 60 7f cb   <.K..F....*..`..
    0010 - 48 80 16 86 ac 60 4f 0f-26 64 e5 73 46 84 74 fb   H....`O.&d.sF.t.

    Start Time: 1694278279
    Timeout   : 7200 (sec)
    Verify return code: 0 (ok)
    Extended master secret: no
    Max Early Data: 0
---
read R BLOCK
```

So basically that's all the problems I've run into so far that I haven't been able to overcome

any thoughts/suggestions appreciated

**Comment by baentsch** [comment]
Thanks for your detailed tests and summary. I cannot comment on what the Cloudflare time server does (is it even PQ enabled??) but I can explain why you are seeing OK interop using x22519_kyber768 and no interop using x25519_kyber512: You ran into a "historical aberration": Cloudflare chose to select another code point than OQS did for that algorithm: From https://github.com/open-quantum-safe/oqs-provider/blob/main/ALGORITHMS.md#changing-code-points:

> the following command can be used to successfully confirm interoperability between the oqs-provider and the Cloudflare infrastructure using this hybrid classic/quantum-safe algorithm:

     OQS_CODEPOINT_X25519_KYBER512=65072  ./openssl/apps/openssl s_client -groups x25519_kyber512 -connect cloudflare.com:443 -provider-path _build/oqsprov -provider oqsprovider -provider default

At the time of introduction there was no standard yet defining the code point for that algorithm, so neither Cloudflare nor OQS does it "wrong" per se. OQS just had an interop test server (for x25519_kyber512) _before_ Cloudflare did and hence did not dare to change the code point we chose before (Cloudflare made their choice) to not destroy interop with old(er) OQS software deployments. For interop testing (before final standards are set), `oqsprovider` has the flexibility to change code points for all algorithms via environment variable(s). You may want to try that out.

In contrast, for "x25519_kyber768", OQS added that (initially from a crypto-strength unusual combination) _after_ Cloudflare, so OQS adapted  to the Cloudflare code point choice. Hence, complete interop "from the start".

@dstebila As we are about to change interop choices soon with `liboqs 0.9.0` what's your take regarding "x25519_kyber512": Is there agreement on that code point? Worthwhile to break interop with our old code to avoid such "usage" surprises?

**Reply by catharsis71** [reply]
>I cannot comment on what the Cloudflare time server does (is it even PQ enabled??) 

They haven't explicitly said, but I still have normal X25519 and a bunch of others enabled, so if they didn't want to use PQ I would expect it to negotiate something else.  In fact if I move normal X25519 to the front of my Group list, ahead of the X25519_kybers, it seems to work again.  So it's like they tried to implement PQ but it didn't work.

I'll try to get in contact with them.

**Comment by dstebila** [comment]
> @dstebila As we are about to change interop choices soon with `liboqs 0.9.0` what's your take regarding "x25519_kyber512": Is there agreement on that code point? Worthwhile to break interop with our old code to avoid such "usage" surprises?

I'm not aware of any plan by the TLS working group to adopt an `x25519_kyber512` codepoint.  So I think we will eventually wind down use of experimental codepoints that did not advance.

**Reply by catharsis71** [reply]
> experimental codepoints that did not advance

So is x25519_kyber768 looking to be the only one that's moving forward?  Or are there others I should test as well?  The full list is pretty daunting, I'd like to focus on at most a few that have a chance of becoming relevant.

**Reply by dstebila** [reply]
Based on my understanding of current intentions in the TLS working group and as documented in https://datatracker.ietf.org/doc/draft-ietf-tls-hybrid-design/, the only two moving forward at this point in time are x25519+kyber768 and secp256r1+kyber768.

**Reply by catharsis71** [reply]
>secp256r1+kyber768

thanks for letting me know about this

but does OQS not support it?

I don't see it on the list, tried it with an underscore and a few other ways, but just get a "bad value" or "failed setting curves list"

**Reply by baentsch** [reply]
> I don't see it on the list, tried it with an underscore and a few other ways, but just get a "bad value" or "failed setting curves list"

Where did you look? In our [algorithm listing](https://github.com/open-quantum-safe/oqs-provider/blob/main/ALGORITHMS.md) it's directly listed right underneath "x25519_kyber768".

**Reply by catharsis71** [reply]
I was looking for secp256 instead of p256 but I get it now

will give it a try

**Comment by catharsis71** [comment]
I think I figured out on my own what's going on with the Cloudflare NTS server

with TLS1.3, the client can only specify in a single curve in the initial Hello message

if the server doesn't like that curve, there's supposed to be a renegotiation

however Cloudflare's NTS server seems to have this renegotiation disabled

they only support X25519, secp256r1, and secp384r1 so the client MUST have one of those three as its first-choice curve

looking further into how ntpsec handles this, you can specify a preferred curve list for incoming NTS SERVER connections, but for NTS client functionality, it seems like you can't configure anything, and it just pulls curve config from /etc/ssl/openssl.cnf

so basically I have to put X25519 first in my openssl.cnf:

```
Groups = x25519:x25519_kyber768:sect571r1:sect571k1:secp521r1:brainpoolP512r1:sect409r1:sect409k1:brainpoolP384r1:secp384r1:X448:ffdhe8192:ffdhe6144:ffdhe4096:ffdhe3072:secp256r1
```

but then I can make ntpsec prefer kyber for incoming connections:

```
nts enable mintls 1.2 tlsecdhcurves x25519_kyber768:sect571r1:sect571k1:secp521r1:brainpoolP512r1:sect409r1:sect409k1:brainpoolP384r1:secp384r1:X448:x25519:ffdhe8192:ffdhe6144:ffdhe4096:ffdhe3072:secp256r1
```

and I can make curl prefer kyber by creating a ```~/.curlrc```

```
--curve x25519_kyber768:sect571r1:sect571k1:secp521r1:brainpoolP512r1:sect409r1:sect409k1:brainpoolP384r1:secp384r1:X448:x25519:ffdhe8192:ffdhe6144:ffdhe4096:ffdhe3072:secp256r1
```

and in my Apache config I make it prefer kyber and drop insecure curves from the list:

```
SSLOpenSSLConfCmd Curves x25519_kyber768:sect571r1:sect571k1:secp521r1:brainpoolP512r1:sect409r1:sect409k1:brainpoolP384r1:secp384r1:X448
```

per prior discussion I've also removed x25519_kyber512 from all of these

**Comment by baentsch** [comment]
> with TLS1.3, the client can only specify in a single curve in the initial Hello message

Where do you have this information from? [The spec](https://datatracker.ietf.org/doc/html/rfc8446#section-4.2.7) clearly documents a "NamedGroupList" (which you then also configure via the various program's different mechanisms to be activated):

> make ntpsec prefer kyber for incoming connections:

> nts enable mintls 1.2 tlsecdhcurves x25519_kyber768:sect571r1:sect571k1:secp521r1:brainpoolP512r1:sect409r1:sect409k1:brainpoolP384r1:secp384r1:X448:x25519:ffdhe8192:ffdhe6144:ffdhe4096:ffdhe3072:secp256r1

This config is key to make any TLS1.3 server prefer the Kyber hybrid as per this statement from the spec:

> ordered from most preferred to least preferred.

So it seems Cloudflare NTS server is PQ enabled and spec compliant (once the groups you want to use are correctly configured). I personally would not allow TLS1.2 as you seem to do, though to ensure only 1.3 (group selection) logic to be used. Alternatively, be sure to set the clients to only permit TLS1.3.

I think the whole discussion can then be closed, right? Please mark as answers what do deem helpful. Hoping you agree with the change of title to allow others to find this in the future.

**Reply by catharsis71** [reply]
> So it seems Cloudflare NTS server is PQ enabled

I don't think it is

sorry if I was unclear in my previous message, I think I was misunderstood

they only support (normal) X25519, secp256r1, and secp384r1, and furthermore, one of those must be the FIRST curve specified by the client

so for example this works (because X25519 is first):

```openssl s_client -connect time.cloudflare.com:4460 -curves x25519:x448```

but this does not work (because X25519 is NOT first):

```openssl s_client -connect time.cloudflare.com:4460 -curves x448:x25519```

this does work (but using normal X25519 instead of x25519_kyber768:

```openssl s_client -connect time.cloudflare.com:4460 -curves x25519:x25519_kyber768```

this does not work (because normal x25519 must be FIRST on the list, not second):

```openssl s_client -connect time.cloudflare.com:4460 -curves x25519_kyber768:x25519```

I will attempt to contact Cloudflare about this

**Comment by baentsch** [comment]
> this does not work (because X25519 is NOT first):
> openssl s_client -connect time.cloudflare.com:4460 -curves x448:x25519

This is extraordinarily weird (IMO out of spec): Even if the server does not support x448 it must be able to fall back to x25519 if the client offers it (and the server apparently supports it).

> I will attempt to contact Cloudflare about this

Please do. Would be curious what the response is (even though it has nothing to do with OQS :)

**Comment by baentsch** [comment]
> >    So it seems Cloudflare NTS server is PQ enabled

> I don't think it is

> sorry if I was unclear in my previous message, I think I was misunderstood

Yes, I did (misunderstand): The statement 

> I can make ntpsec prefer kyber for incoming connections:

> nts enable mintls 1.2 tlsecdhcurves x25519_kyber768:sect571r1:sect571k1:secp521r1:brainpoolP512r1:sect409r1:sect409k1:brainpoolP384r1:secp384r1:X448:x25519:ffdhe8192:ffdhe6144:ffdhe4096:ffdhe3072:secp256r1

led me to believe it worked. But looking at the [Cloudflare blog](https://blog.cloudflare.com/secure-time/) this server seems to pre-date their PQ activities -- so most likely it indeed doesn't support those algs (and even TLS1.3 works "strangely").

**Comment by beldmit** [comment]
I have a vague memory that openssl sends only the first group provided by `curves ` but it should be double checked

**Reply by baentsch** [reply]
Yikes -- that would make the implementation awfully inefficient, wouldn't it? Should be easy to check via `wireshark`...

**Reply by baentsch** [reply]
Tracing the result of `openssl s_client -tls1_3 -connect time.cloudflare.com:4460 -curves x25519:x448`: 

![grafik](https://github.com/open-quantum-safe/oqs-provider/assets/57787676/e2b9a06d-be99-4b39-8fc4-7be19a16aad6)

--> Looks like openssl does send both groups in the client hello

**Reply by catharsis71** [reply]
>I have a vague memory that openssl sends only the first group provided by curves 

that matches what I found on a couple websites this morning after hours of searching but I'm having trouble finding them again

>Should be easy to check via wireshark

would a tcpdump capture work?

Ran this capture (using /etc/hosts temporarily to force a consistent IP for time.cloudflare.com):  
```tcpdump -w - -U host 162.159.200.123 > capture```

Ran this command with this output (failed to connect because x25519 is 2nd item on the curve list instead of the 1st):

```
# openssl s_client -connect time.cloudflare.com:4460 -curves x25519_kyber768:x25519
CONNECTED(00000003)
40B7F974077F0000:error:0A0003E7:SSL routines:tls_process_server_hello:invalid session id:../ssl/statem/statem_clnt.c:1467:
---
no peer certificate available
---
No client certificate CA names sent
---
SSL handshake has read 61 bytes and written 1496 bytes
Verification: OK
---
New, (NONE), Cipher is (NONE)
Secure Renegotiation IS NOT supported
Compression: NONE
Expansion: NONE
No ALPN negotiated
Early data was not sent
Verify return code: 0 (ok)
---
```

The capture file is binary so I uuencoded it:  https://pastebin.com/LuVyGfcx

**Reply by beldmit** [reply]
I'll double check what I mixed it with. Sorry for the false alarm

**Reply by beldmit** [reply]
Probably I mixed it with the `key_share` extension

**Reply by catharsis71** [reply]
I think key_share is what I found information on this morning

Is it possible that the Cloudflare NTS server only looks at key_share and disregards supported_groups?

**Reply by baentsch** [reply]
> Is it possible that the Cloudflare NTS server only looks at key_share and disregards supported_groups?

That wouldn't be exactly spec compliant, would it be? It would be interesting to learn which library provides their TLS stack. I didn't find mention of this in their blog on NTS.

**Reply by catharsis71** [reply]
> It would be interesting to learn which library provides their TLS stack

According to https://blog.cloudflare.com/announcing-cfnts/ they're using rustls

I opened a bug on https://github.com/cloudflare/cfnts which the software they run on time.cloudflare.com

if that doesn't go anywhere I'll try to contact them some other way

if time allows I'll also try to set up my own cfnts server to see if it behaves the same way as theirs

**Comment by baentsch** [comment]
@catharsis71 Can we consider this resolved/answered for oqs-provider? Seems like the action now is at https://github.com/cloudflare/cfnts/issues/64
